File: thai_tax/thai_tax/report/purchase_tax_report/purchase_tax_report.js
// Copyright (c) 2023, Kitti U. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Purchase Tax Report"] = {
	filters: [
		{
			"fieldname":"filter_based_on",
			"label": __("Filter Based On"),
			"fieldtype": "Select",
			"options": ["Fiscal Year", "Date Range"],
			"default": ["Fiscal Year"],
			"reqd": 1,
			on_change: function() {
				let filter_based_on = frappe.query_report.get_filter_value('filter_based_on');
				frappe.query_report.toggle_filter_display('year', filter_based_on === 'Date Range');
				frappe.query_report.toggle_filter_display('month', filter_based_on === 'Date Range');
				frappe.query_report.toggle_filter_display('start_date', filter_based_on === 'Fiscal Year');
				frappe.query_report.toggle_filter_display('end_date', filter_based_on === 'Fiscal Year');

				frappe.query_report.refresh();
			}
		},
		{
			fieldname: "year",
			label: __("Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
		},
		{
			fieldname: "month",
			label: __("Month"),
			fieldtype: "Select",
			options: ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12"],
		},
		{
			fieldname: "start_date",
			label: __("Start Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "end_date",
			label: __("End Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "company_tax_address",
			label: __("Company Address"),
			fieldtype: "Link",
			options: "Address",
			get_query: () => {
				return {
					filters: {
						is_your_company_address: 1,
					},
				};
			},
		},
	],
};

File: thai_tax/thai_tax/report/purchase_tax_report/purchase_tax_report.py
# Copyright (c) 2023, Kitti U. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.query_builder import CustomFunction


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data, None, None, None


def get_columns():
	return [
		{
			"label": _("Tax Address"),
			"fieldname": "company_tax_address",
			"fieldtype": "Link",
			"options": "Address",
			"width": 0,
		},
		{
			"label": _("Report Date"),
			"fieldname": "report_date",
			"fieldtype": "Date",
			"width": 0,
		},
		{
			"label": _("Number"),
			"fieldname": "name",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Supplier"),
			"fieldname": "party_name",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Tax ID"),
			"fieldname": "tax_id",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Branch"),
			"fieldname": "branch_code",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Supplier Address"),
			"fieldname": "supplier_address",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Tax Base"),
			"fieldname": "tax_base",
			"fieldtype": "Currency",
			"options": "Company:company:default_currency",
			"width": 0,
		},
		{
			"label": _("Tax Amount"),
			"fieldname": "tax_amount",
			"fieldtype": "Currency",
			"options": "Company:company:default_currency",
			"width": 0,
		},
		{
			"label": _("Ref Voucher Type"),
			"fieldname": "voucher_type",
			"fieldtype": "Data",
			"width": 0,
		},
		{
			"label": _("Ref Voucher No"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "voucher_type",
			"width": 0,
		},
		{
			"label": _("Ref Tax Invoice"),
			"fieldname": "tax_invoice",
			"fieldtype": "Link",
			"options": "Purchase Tax Invoice",
			"width": 0,
		},
	]


def get_data(filters):

	tinv = frappe.qb.DocType("Purchase Tax Invoice")
	sup = frappe.qb.DocType("Supplier")
	addr = frappe.qb.DocType("Address")
	round = CustomFunction("round", ["value", "digit"])
	coalesce = CustomFunction("coalesce", ["value1", "value2"])
	month = CustomFunction("month", ["date"])
	year = CustomFunction("year", ["date"])
	concat_ws = CustomFunction("concat_ws", ["separator", "1", "2", "3", "4", "5", "6"])

	query = (
		frappe.qb.from_(tinv)
		.left_join(sup)
		.on(sup.name == tinv.party)
		.left_join(addr)
		.on(addr.name == sup.supplier_primary_address)
		.select(
			tinv.company_tax_address.as_("company_tax_address"),
			tinv.report_date.as_("report_date"),
			coalesce(tinv.number, tinv.name).as_("name"),
			sup.supplier_name.as_("party_name"),
			sup.tax_id.as_("tax_id"),
			sup.branch_code.as_("branch_code"),
			concat_ws(
				" ",
				addr.address_line1,
				addr.address_line2,
				addr.city,
				addr.county,
				addr.state,
				addr.pincode,
			).as_("supplier_address"),
			round(tinv.tax_base, 2).as_("tax_base"),
			round(tinv.tax_amount, 2).as_("tax_amount"),
			tinv.voucher_type.as_("voucher_type"),
			tinv.voucher_no.as_("voucher_no"),
			tinv.name.as_("tax_invoice"),
		)
		.where(
			(tinv.docstatus == 1)
			& (month(tinv.report_date) == filters.get("month"))
			& (year(tinv.report_date) == filters.get("year"))
		)
		.orderby(tinv.report_date)
	)

	if filters.get("company_tax_address"):
		query = query.where(tinv.company_tax_address == filters.get("company_tax_address"))

	result = query.run(as_dict=True)

	return result
